File: Tokenizer_C#/TokenizerLib/TikTokenizer.cs
﻿// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using Microsoft.DeepDev.TokenizerLib.Utils;

namespace Microsoft.DeepDev
{
    /// <summary>
    /// This is a C# version of OpenAI's tiktoken implementation of
    /// Byte pair encoding(BPE): https://en.wikipedia.org/wiki/Byte_pair_encoding,
    /// the goal is to support context tokenization for OpenAI large language models
    /// in .NET runtime.
    /// Reference: https://github.com/openai/tiktoken/blob/main/src/lib.rs
    /// </summary>
    public class TikTokenizer : ITokenizer
    {

        private IReadOnlyDictionary<string, int> SpecialTokensEncoder = null!;
        private Regex Regex = null!;
        private IReadOnlyDictionary<byte[], int> Encoder = null!;
        private IReadOnlyDictionary<int, byte[]> Decoder = null!;
        private Regex SpecialTokensRegex = null!;
        private IReadOnlyDictionary<int, string> SpecialTokensDecoder = null!;

        /// <summary>
        ///     The default LRU cache size.
        /// </summary>
        public const int DefaultCacheSize = 4096;

        private readonly LruCache<string, int[]> Cache;

        public int NumOfCacheEntries => this.Cache.Count;

        /// <summary>
        /// Take the encoder tokens mapping from OpenAI tiktoken dump to build the encoder
        /// For gpt-3.5-turbo/gpt4, you can download the BPE tokens mapping from:
        /// https://openaipublic.blob.core.windows.net/encodings/cl100k_base.tiktoken
        /// </summary>
        /// <param name="encoder">Encoder mapping</param>
        /// <param name="specialTokensEncoder">Handle special tokens</param>
        /// <param name="pattern">Regex pattern to break a string to be encoded</param>
        public TikTokenizer(IReadOnlyDictionary<byte[], int> encoder, IReadOnlyDictionary<string, int> specialTokensEncoder, string pattern, int cacheSize = DefaultCacheSize)
        {
            Cache = new LruCache<string, int[]>(cacheSize);
            Init(encoder, specialTokensEncoder, pattern);
        }

        /// <summary>
        /// Overload of the constructor above to load encoder from a file stream
        /// </summary>
        /// <param name="tikTokenBpeFileStream">Encoder mapping file stream</param>
        /// <param name="specialTokensEncoder">Handle special tokens</param>
        /// <param name="pattern">Regex pattern to break a string to be encoded</param>
        public TikTokenizer(Stream tikTokenBpeFileStream, IReadOnlyDictionary<string, int> specialTokensEncoder, string pattern, int cacheSize = DefaultCacheSize)
        {
            Cache = new LruCache<string, int[]>(cacheSize);
            var encoder = LoadTikTokenBpe(tikTokenBpeFileStream);
            Init(encoder, specialTokensEncoder, pattern);
        }

        /// <summary>
        /// Initialize encoders and decoders
        /// </summary>
        /// <param name="encoder">Map byte[] to integer token id</param>
        /// <param name="specialTokensEncoder">Map special token string to integer token id</param>
        /// <param name="pattern">Regex pattern to break a string into pieces before encoding</param>
        /// <exception cref="ArgumentException">Throws if encoder and decoder entries are not matching</exception>
        private void Init(IReadOnlyDictionary<byte[], int> encoder, IReadOnlyDictionary<string, int> specialTokensEncoder, string pattern)
        {
            Encoder = encoder;
            Regex = new Regex(pattern, RegexOptions.Compiled);
            SpecialTokensRegex = new Regex(string.Join("|", specialTokensEncoder.Keys.Select(s => Regex.Escape(s))), RegexOptions.Compiled);
            SpecialTokensEncoder = specialTokensEncoder;

            Decoder = Encoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);

            if (Encoder.Count != Decoder.Count)
            {
                throw new ArgumentException("Encoder and decoder sizes don't match");
            }

            SpecialTokensDecoder = specialTokensEncoder.ToDictionary(kvp => kvp.Value, kvp => kvp.Key);

        }

        /// <summary>
        /// Load BPE rank dictionary from a stream.
        /// </summary>
        /// <param name="tikTokenBpeFileStream">Stream to the BPE rank file</param>
        /// <returns>Map of byte[] to integer token id</returns>
        /// <exception cref="InvalidOperationException"></exception>
        private Dictionary<byte[], int> LoadTikTokenBpe(Stream tikTokenBpeFileStream)
        {
            var bpeDict = new Dictionary<byte[], int>(new ByteArrayComparer());
            try
            {
                using (StreamReader reader = new StreamReader(tikTokenBpeFileStream))
                {
                    while (!reader.EndOfStream)
                    {
                        string line = reader.ReadLine();
                        if (string.IsNullOrWhiteSpace(line))
                        {
                            continue;
                        }

                        var tokens = line.Split(' ');
                        if (tokens.Length != 2)
                        {
                            throw new FormatException($"Invalid format in the BPE encoder file stream");
                        }

                        var tokenBytes = Convert.FromBase64String(tokens[0]);
                        int rank = 0;
                        if (int.TryParse(tokens[1], out rank))
                        {
                            bpeDict[tokenBytes] = rank;
                        }
                        else
                        {
                            throw new FormatException($"Can't parse {tokens[1]} to integer");
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Failed to load from BPE encoder file stream: {ex.Message}", ex);
            }

            return bpeDict;
        }

        /// <summary>
        /// Encode a string with a set of allowed special tokens that are not broken apart.
        /// </summary>
        /// <param name="text">String to be encoded</param>
        /// <param name="allowedSpecial">A set of special tokens could appear in the text</param>
        /// <returns>List of token ids</returns>
        public List<int> Encode(string text, IReadOnlyCollection<string> allowedSpecial)
        {
            var tokenIds = new List<int>();
            int start = 0;
            while (true)
            {
                Match nextSpecial;
                int end;
                FindNextSpecialToken(text, allowedSpecial, start, out nextSpecial, out end);
                if (end > start)
                {
                    Encode(text, tokenIds, start, end);
                }

                if (nextSpecial.Success)
                {
                    start = EncodeSpecialToken(tokenIds, nextSpecial);
                    if (start >= text.Length)
                    {
                        break;
                    }
                }
                else
                {
                    break;
                }
            }

            return tokenIds;
        }

        /// <summary>
        /// Encode a special token matched through regex.
        /// </summary>
        /// <param name="tokenIds">The list of token ids to add the special token id</param>
        /// <param name="nextSpecial">Regex match of the special token</param>
        /// <returns>The start index for next special token search</returns>
        private int EncodeSpecialToken(List<int> tokenIds, Match nextSpecial)
        {
            var token = SpecialTokensEncoder[nextSpecial.Value];
            tokenIds.Add(token);
            return nextSpecial.Index + nextSpecial.Length;
        }

        /// <summary>
        /// Search for special token in a string
        /// </summary>
        /// <param name="text">String to be searched</param>
        /// <param name="allowedSpecial">Allowed special tokens</param>
        /// <param name="start">Start search index in the string</param>
        /// <param name="nextSpecial">The regex match of a special token</param>
        /// <param name="end">The index of the special token matched or the end of the text</param>
        private void FindNextSpecialToken(string text, IReadOnlyCollection<string> allowedSpecial, int start, out Match nextSpecial, out int end)
        {
            int startFind = start;
            while (true)
            {
                nextSpecial = SpecialTokensRegex.Match(text, startFind);
                if (!nextSpecial.Success) break;
                if (allowedSpecial.Contains(text.Substring(nextSpecial.Index, nextSpecial.Length))) break;
                startFind = nextSpecial.Index + 1;
            }
            end = nextSpecial.Success ? nextSpecial.Index : text.Length;
        }

        /// <summary>
        /// Encode a string based between start and end index
        /// </summary>
        /// <param name="text">The original string to be encoded</param>
        /// <param name="tokenIds">The list of token ids to add encoded token id</param>
        /// <param name="start">Start index</param>
        /// <param name="end">End index</param>
        private void Encode(string text, List<int> tokenIds, int start, int end)
        {
            foreach (Match match in Regex.Matches(text[start..end]))
            {
                if (this.Cache.Lookup(match.Value, out int[] tokens))
                {
                    tokenIds.AddRange(tokens);
                }
                else
                {
                    //cache miss
                    var bytes = Encoding.UTF8.GetBytes(match.Value);
                    if (Encoder.TryGetValue(bytes, out int token))
                    {
                        tokenIds.Add(token);
                    }
                    else
                    {
                        var encodedTokens = BytePairEncoder.BytePairEncode(bytes, Encoder);
                        tokenIds.AddRange(encodedTokens);
                        this.Cache.Add(match.Value, encodedTokens.ToArray());
                    }
                }
            }
        }

        /// <summary>
        /// Encode a string from start index to end index based on max token count,
        /// trim suffix if the token count is greater than max token count
        /// </summary>
        /// <param name="text">The original string to be encoded</param>
        /// <param name="tokenIds">The list of token ids to add encoded token id</param>
        /// <param name="start">Start index</param>
        /// <param name="end">End index</param>
        /// <param name="maxTokenCount">Max token count</param>
        /// <param name="tokenCount">The token count before the start index</param>
        /// <param name="encodeLength">The encoded string length before the start index</param>
        /// <returns>The new token count and encoded string length</returns>
        private (int TokenCount, int EncodeLength) EncodeTrimSuffix(string text, List<int> tokenIds, int start, int end, int maxTokenCount, int tokenCount, int encodeLength)
        {
            foreach (Match match in Regex.Matches(text[start..end]))
            {
                var piece = match.Value;
                if (this.Cache.Lookup(piece, out int[] tokens))
                {
                    tokenCount += tokens.Length;
                    if (tokenCount <= maxTokenCount)
                    {
                        encodeLength += piece.Length;
                        tokenIds.AddRange(tokens);
                    }
                    else
                    {
                        break;
                    }
                }
                else
                {
                    //cache miss
                    var bytes = Encoding.UTF8.GetBytes(piece);
                    if (Encoder.TryGetValue(bytes, out int token))
                    {
                        tokenCount++;
                        if (tokenCount <= maxTokenCount)
                        {
                            encodeLength += piece.Length;
                            tokenIds.Add(token);
                        }
                        else
                        {
                            break;
                        }
                    }
                    else
                    {
                        var encodedTokens = BytePairEncoder.BytePairEncode(bytes, Encoder);
                        this.Cache.Add(piece, encodedTokens.ToArray());
                        tokenCount += encodedTokens.Count;
                        if (tokenCount <= maxTokenCount)
                        {
                            encodeLength += piece.Length;
                            tokenIds.AddRange(encodedTokens);
                        }
                        else
                        {
                            break;
                        }
                    }
                }
                if (tokenCount >= maxTokenCount) break;
            }
            return (tokenCount, encodeLength);
        }

        /// <summary>
        /// Encode a piece of text limited by max token count through trimming suffix
        /// </summary>
        /// <param name="text">Text to be encoded</param>
        /// <param name="allowedSpecial">A set of special tokens could appear in the text</param>
        /// <param name="maxTokenCount">The max token count</param>
        /// <returns>(List<int> TokenIds, string Text) - Token ids and text after suffix truncation based on max token count</returns>
        public (List<int> TokenIds, string Text) EncodeTrimSuffix(string text, IReadOnlyCollection<string> allowedSpecial, int maxTokenCount)
        {
            var tokenIds = new List<int>();

            int start = 0;
            int tokenCount = 0;
            var encodeLength = 0;
            while (true)
            {
                Match nextSpecial;
                int end;
                FindNextSpecialToken(text, allowedSpecial, start, out nextSpecial, out end);

                if (end > start)
                {
                    (tokenCount, encodeLength) = EncodeTrimSuffix(text, tokenIds, start, end, maxTokenCount, tokenCount, encodeLength);

                    if (tokenCount >= maxTokenCount)
                    {
                        break;
                    }
                }

                if (nextSpecial.Success)
                {
                    tokenCount++;
                    if (tokenCount <= maxTokenCount)
                    {
                        start = EncodeSpecialToken(tokenIds, nextSpecial);
                        encodeLength += nextSpecial.Value.Length;
                        if (start >= text.Length)
                        {
                            break;
                        }
                    }
                    if (tokenCount >= maxTokenCount)
                    {
                        break;
                    }
                }
                else
                {
                    break;
                }
            }

            var encodedText = encodeLength == text.Length ? text : text[..encodeLength];

            return (tokenIds, encodedText);
        }

        /// <summary>
        /// Encode a piece of text limited by max token count through trimming prefix
        /// </summary>
        /// <param name="text">Text to be encoded</param>
        /// <param name="allowedSpecial">A set of special tokens could appear in the text</param>
        /// <param name="maxTokenCount">The max token count</param>
        /// <returns>(List<int> TokenIds, string Text) - Token ids and text after prefix truncation based on max token count</returns>
        public (List<int> TokenIds, string Text) EncodeTrimPrefix(string text, IReadOnlyCollection<string> allowedSpecial, int maxTokenCount)
        {
            var tokenIds = new List<int>();

            int start = 0;
            int tokenCount = 0;
            var encodeLength = 0;
            var tokenCountMap = new SortedDictionary<int, int>();
            tokenCountMap.Add(tokenCount, encodeLength);
            while (true)
            {
                Match nextSpecial;
                int end;
                FindNextSpecialToken(text, allowedSpecial, start, out nextSpecial, out end);

                if (end > start)
                {
                    foreach (Match match in Regex.Matches(text[start..end]))
                    {
                        var piece = match.Value;

                        if (this.Cache.Lookup(match.Value, out int[] tokens))
                        {
                            tokenCount += tokens.Length;
                            encodeLength += piece.Length;
                            tokenIds.AddRange(tokens);
                            tokenCountMap[tokenCount] = encodeLength;
                        }
                        else
                        {
                            var bytes = Encoding.UTF8.GetBytes(piece);
                            if (Encoder.TryGetValue(bytes, out int token))
                            {
                                tokenCount++;
                                encodeLength += piece.Length;
                                tokenIds.Add(token);
                                tokenCountMap[tokenCount] = encodeLength;

                            }
                            else
                            {
                                var encodedTokens = BytePairEncoder.BytePairEncode(bytes, Encoder);
                                this.Cache.Add(piece, encodedTokens.ToArray());
                                tokenCount += encodedTokens.Count;
                                encodeLength += piece.Length;
                                tokenIds.AddRange(encodedTokens);
                                tokenCountMap[tokenCount] = encodeLength;
                            }
                        }
                    }
                }

                if (nextSpecial.Success)
                {
                    start = EncodeSpecialToken(tokenIds, nextSpecial);
                    tokenCount++;
                    encodeLength += nextSpecial.Value.Length;
                    tokenCountMap[tokenCount] = encodeLength;
                    if (start >= text.Length)
                    {
                        break;
                    }
                }
                else
                {
                    break;
                }
            }

            if (tokenCount <= maxTokenCount)
            {
                return (tokenIds, text);
            }

            var prefixTokenCount = tokenCount - maxTokenCount;
            var actualPrefixTokenCount = 0;
            var actualPrefixStrLength = 0;
            foreach (var pair in tokenCountMap)
            {
                if (pair.Key >= prefixTokenCount)
                {
                    actualPrefixTokenCount = pair.Key;
                    actualPrefixStrLength = pair.Value;
                    break;
                }
            }

            return (tokenIds.Skip(actualPrefixTokenCount).ToList(), text[actualPrefixStrLength..]);
        }

        /// <summary>
        /// Decode an array of integer token ids
        /// </summary>
        /// <param name="tokens">An array of integer token ids</param>
        /// <returns>Decoded text</returns>
        public string Decode(int[] tokens)
        {
            var decoded = new List<byte>(tokens.Length * 2);
            foreach (var token in tokens)
            {
                byte[] tokenBytes = { };
                if (Decoder.TryGetValue(token, out var value))
                {
                    tokenBytes = value;
                }
                else if (SpecialTokensDecoder.TryGetValue(token, out var specialTokenValue))
                {
                    tokenBytes = Encoding.UTF8.GetBytes(specialTokenValue);
                }
                decoded.AddRange(tokenBytes);
            }

            return Encoding.UTF8.GetString(decoded.ToArray());
        }
    }

}

File: tokenizer_ts/src/tikTokenizer.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import * as fs from "fs";
import { LRUCache } from "lru-cache";
import { TextDecoder } from "util";
import { BinaryMap, bytePairEncode } from "./bytePairEncode";
import { makeTextEncoder } from './textEncoder';

/**
 * Load BPE ranks from a file
 * @param tikTokenBpeFile BPE file path
 * @returns Map<Uint8Array, number> BPE ranks
 */
function loadTikTokenBpe(tikTokenBpeFile: string): Map<Uint8Array, number> {
  const bpeDict = new Map<Uint8Array, number>();
  try {
    const fileContent = fs.readFileSync(tikTokenBpeFile, "utf-8");
    processBpeRanks(fileContent);
    return bpeDict;
  } catch (ex) {
    throw new Error(`Failed to load from BPE encoder file stream: ${ex}`);
  }

  function processBpeRanks(fileContent: string) {
    for (const line of fileContent.split(/[\r\n]+/)) {
      if (line.trim() === "") {
        continue;
      }

      const tokens = line.split(" ");
      if (tokens.length !== 2) {
        throw new Error("Invalid format in the BPE encoder file stream");
      }

      const tokenBytes = new Uint8Array(Buffer.from(tokens[0], "base64"));
      const rank = parseInt(tokens[1]);
      if (!isNaN(rank)) {
        bpeDict.set(tokenBytes, rank);
      } else {
        throw new Error(`Can't parse ${tokens[1]} to integer`);
      }
    }
  }
}

/**
 * Escape special characters in a regex string
 * @param regex regex string
 */
function escapeRegExp(regex: string) {
  return regex.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * This is a Typescript version of OpenAI's tiktoken implementation of
 * Byte pair encoding(BPE): https://en.wikipedia.org/wiki/Byte_pair_encoding,
 * the goal is to support context tokenization for OpenAI large language models in .NET runtime.
 * Reference: https://github.com/openai/tiktoken/blob/main/src/lib.rs
 */
export class TikTokenizer {
  private regex?: RegExp;
  private encoder?: BinaryMap<number>;
  private decoder?: Map<number, Uint8Array>;
  private specialTokensRegex?: RegExp;
  private specialTokensEncoder?: ReadonlyMap<string, number>;
  private specialTokensDecoder?: Map<number, string>;
  private textEncoder = makeTextEncoder();
  private textDecoder = new TextDecoder("utf-8");
  public readonly cache: LRUCache<string, number[]>;

  /**
   * Take the encoder tokens mapping from OpenAI tiktoken dump to build the encoder
   * For gpt-3.5-turbo/gpt4, you can download the BPE tokens mapping from:
   * https://openaipublic.blob.core.windows.net/encodings/cl100k_base.tiktoken
   * @param tikTokenBpeFile BPE rank file path
   * @param specialTokensEncoder special tokens encoder
   * @param regexPattern regex pattern to split the input text
   * @param cacheSize cache size
   */
  constructor(
    tikTokenBpeFile: string,
    specialTokensEncoder: ReadonlyMap<string, number>,
    regexPattern: string,
    cacheSize: number = 8192
  ) {
    const options = { max: cacheSize };
    this.cache = new LRUCache(options);
    const bpeDict = loadTikTokenBpe(tikTokenBpeFile);
    this.init(bpeDict, specialTokensEncoder, regexPattern);
  }

  protected init(
    bpeDict: ReadonlyMap<Uint8Array, number>,
    specialTokensEncoder: ReadonlyMap<string, number>,
    regexPattern: string
  ): void {
    this.encoder = new BinaryMap();
    for (const [key, value] of bpeDict) {
      this.encoder.set(key, value);
    }
    this.regex = new RegExp(regexPattern, "gu");
    this.specialTokensRegex = new RegExp(
      Array.from(specialTokensEncoder.keys())
        .map(s => escapeRegExp(s))
        .join("|")
    );
    this.specialTokensEncoder = specialTokensEncoder;

    this.decoder = new Map<number, Uint8Array>();
    for (const [key, value] of bpeDict) {
      this.decoder.set(value, key);
    }

    if (bpeDict.size !== this.decoder.size) {
      throw new Error("Encoder and decoder sizes do not match");
    }

    this.specialTokensDecoder = new Map<number, string>();
    for (const [key, value] of specialTokensEncoder) {
      this.specialTokensDecoder.set(value, key);
    }
  }

  private findNextSpecialToken(
    text: string,
    start: number,
    allowedSpecial?: ReadonlyArray<string>
  ): [RegExpMatchArray | null, number] {
    let startFind = start;
    let nextSpecial: RegExpMatchArray | null = null;
    if (allowedSpecial && this.specialTokensRegex) {
      while (true) {
        nextSpecial = text.slice(startFind).match(this.specialTokensRegex);
        if (!nextSpecial) {
          break;
        }
        if (allowedSpecial && allowedSpecial.includes(nextSpecial[0])) {
          break;
        }
        startFind += nextSpecial.index! + 1;
      }
    }
    const end = nextSpecial ? startFind + nextSpecial.index! : text.length;
    return [nextSpecial, end];
  }

  /**
   * Encode a string with a set of allowed special tokens that are not broken apart.
   * @param text text to encode
   * @param allowedSpecial allowed special tokens
   * @returns number[] encoded token ids
   */
  public encode(
    text: string,
    allowedSpecial?: ReadonlyArray<string>
  ): number[] {
    const tokenIds: number[] = [];
    let start = 0;
    while (true) {
      let nextSpecial: RegExpMatchArray | null;
      let end: number;
      [nextSpecial, end] = this.findNextSpecialToken(
        text,
        start,
        allowedSpecial
      );
      if (end > start) {
        this.encodeByIndex(text, tokenIds, start, end);
      }

      if (nextSpecial) {
        start = start + this.encodeSpecialToken(tokenIds, nextSpecial);
        if (start >= text.length) {
          break;
        }
      } else {
        break;
      }
    }

    return tokenIds;
  }

  private encodeSpecialToken(
    tokenIds: number[],
    nextSpecial: RegExpMatchArray
  ): number {
    const token = this.specialTokensEncoder?.get(nextSpecial[0]);
    tokenIds.push(token!);
    return nextSpecial.index! + nextSpecial[0].length;
  }

  private encodeByIndex(
    text: string,
    tokenIds: number[],
    start: number,
    end: number
  ): void {
    let match: RegExpExecArray | null;
    const substring = text.substring(start, end);
    this.regex!.lastIndex = 0;
    while ((match = this.regex!.exec(substring))) {
      const cached = this.cache.get(match[0]);
      if (cached) {
        tokenIds.push(...cached);
      } else {
        // cache miss
        const bytes = this.textEncoder.encode(match[0]);
        const token = this.encoder?.get(bytes, 0, this.textEncoder.length);
        if (token !== undefined) {
          tokenIds.push(token);
          this.cache.set(match[0], [token]);
        } else {
          const encodedTokens = bytePairEncode(bytes, this.encoder!, this.textEncoder.length);
          tokenIds.push(...encodedTokens);
          this.cache.set(match[0], encodedTokens);
        }
      }
    }
  }

  private encodeTrimSuffixByIndex(
    text: string,
    tokenIds: number[],
    start: number,
    end: number,
    maxTokenCount: number,
    tokenCount: number,
    encodeLength: number
  ): { tokenCount: number; encodeLength: number } {
    let match: RegExpExecArray | null;
    const substring = text.substring(start, end);
    this.regex!.lastIndex = 0;
    while ((match = this.regex!.exec(substring))) {
      const piece = match[0];
      if (this.cache.has(piece)) {
        let tokens = this.cache.get(piece);
        if (tokenCount + tokens!.length <= maxTokenCount) {
          tokenCount += tokens!.length;
          encodeLength += piece.length;
          tokenIds.push(...tokens!);
        } else {
          let remainingTokens = maxTokenCount - tokenCount;
          tokenCount += remainingTokens;
          encodeLength += piece.length;
          tokenIds.push(...tokens!.slice(0, remainingTokens));
          break;
        }
      } else {
        // cache miss
        const bytes = this.textEncoder.encode(piece);
        const token = this.encoder!.get(bytes, 0, bytes.length);
        if (token !== undefined) {
          this.cache.set(piece, [token]);
          if (tokenCount + 1 <= maxTokenCount) {
            tokenCount++;
            encodeLength += piece.length;
            tokenIds.push(token);
          } else {
            break;
          }
        } else {
          const encodedTokens = bytePairEncode(bytes, this.encoder!, this.textEncoder.length);
          this.cache.set(piece, encodedTokens);
          if (tokenCount + encodedTokens.length <= maxTokenCount) {
            tokenCount += encodedTokens.length;
            encodeLength += piece.length;
            tokenIds.push(...encodedTokens);
          } else {
            let remainingTokens = maxTokenCount - tokenCount;
            tokenCount += remainingTokens;
            encodeLength += piece.length;
            tokenIds.push(...encodedTokens.slice(0, remainingTokens));
            break;
          }
        }
      }
      if (tokenCount >= maxTokenCount) {
        break;
      }
    }

    return { tokenCount, encodeLength };
  }

  /**
   * Encode a piece of text limited by max token count through trimming suffix
   * @param text text to encode
   * @param maxTokenCount max token count
   * @param allowedSpecial allowed special tokens
   * @returns { tokenIds: number[], text: string } encoded token ids and trimmed text
   */
  public encodeTrimSuffix(
    text: string,
    maxTokenCount: number,
    allowedSpecial: ReadonlyArray<string>
  ): { tokenIds: number[]; text: string } {
    const tokenIds: number[] = [];

    let start = 0;
    let tokenCount = 0;
    let encodeLength = 0;
    while (true) {
      let nextSpecial: RegExpMatchArray | null;
      let end: number;
      [nextSpecial, end] = this.findNextSpecialToken(
        text,
        start,
        allowedSpecial
      );

      if (end > start) {
        const {
          tokenCount: newTokenCount,
          encodeLength: newEncodeLength
        } = this.encodeTrimSuffixByIndex(
          text,
          tokenIds,
          start,
          end,
          maxTokenCount,
          tokenCount,
          encodeLength
        );
        tokenCount = newTokenCount;
        encodeLength = newEncodeLength;

        if (tokenCount >= maxTokenCount) {
          break;
        }
      }

      if (nextSpecial !== null) {
        tokenCount++;
        if (tokenCount <= maxTokenCount) {
          start = start + this.encodeSpecialToken(tokenIds, nextSpecial!);
          encodeLength += nextSpecial![0].length;
          if (start >= text.length) {
            break;
          }
        }
        if (tokenCount >= maxTokenCount) {
          break;
        }
      } else {
        break;
      }
    }

    const encodedText =
      encodeLength === text.length ? text : text.slice(0, encodeLength);

    return { tokenIds, text: encodedText };
  }

  /**
   * Encode a piece of text limited by max token count through trimming prefix
   * @param text text to encode
   * @param maxTokenCount max token count
   * @param allowedSpecial allowed special tokens
   * @returns { tokenIds: number[], text: string } encoded token ids and trimmed text
   */
  public encodeTrimPrefix(
    text: string,
    maxTokenCount: number,
    allowedSpecial?: ReadonlyArray<string>
  ): { tokenIds: number[]; text: string } {
    const tokenIds: number[] = [];

    let start = 0;
    let tokenCount = 0;
    let encodeLength = 0;
    const tokenCountMap = new Map<number, number>();
    tokenCountMap.set(tokenCount, encodeLength);
    while (true) {
      let nextSpecial: RegExpMatchArray | null;
      let end: number;
      [nextSpecial, end] = this.findNextSpecialToken(
        text,
        start,
        allowedSpecial
      );

      if (end > start) {
        let match: RegExpExecArray | null;
        const substring = text.substring(start, end);
        this.regex!.lastIndex = 0;
        while ((match = this.regex!.exec(substring))) {
          const piece = match[0];

          if (this.cache.has(piece)) {
            let tokens = this.cache.get(piece);
            tokenCount += tokens!.length;
            encodeLength += piece.length;
            tokenIds.push(...tokens!);
            tokenCountMap.set(tokenCount, encodeLength);
          } else {
            const bytes = this.textEncoder.encode(piece);
            const token = this.encoder!.get(bytes);
            if (token !== undefined) {
              this.cache.set(piece, [token]);
              tokenCount++;
              encodeLength += piece.length;
              tokenIds.push(token);
              tokenCountMap.set(tokenCount, encodeLength);
            } else {
              const encodedTokens = bytePairEncode(bytes, this.encoder!, this.textEncoder.length);
              this.cache.set(piece, encodedTokens);
              tokenCount += encodedTokens.length;
              encodeLength += piece.length;
              tokenIds.push(...encodedTokens);
              tokenCountMap.set(tokenCount, encodeLength);
            }
          }
        }
      }

      if (nextSpecial !== null) {
        start = start + this.encodeSpecialToken(tokenIds, nextSpecial);
        tokenCount++;
        encodeLength += nextSpecial[0].length;
        tokenCountMap.set(tokenCount, encodeLength);
        if (start >= text.length) {
          break;
        }
      } else {
        break;
      }
    }

    if (tokenCount <= maxTokenCount) {
      return { tokenIds, text };
    }

    const prefixTokenCount = tokenCount - maxTokenCount;
    let actualPrefixTokenCount = 0;
    let actualPrefixStrLength = 0;
    for (const [key, value] of tokenCountMap) {
      if (key >= prefixTokenCount) {
        actualPrefixTokenCount = key;
        actualPrefixStrLength = value;
        break;
      }
    }

    // Naive approach if chunks are incorrect
    if (actualPrefixTokenCount > maxTokenCount) {
      const encodedTokens = this.encode(text, allowedSpecial);
      const slicedTokens = encodedTokens.slice(encodedTokens.length - maxTokenCount);
      return {
        tokenIds: slicedTokens,
        text: this.decode(slicedTokens)
      };
    }

    return {
      tokenIds: tokenIds.slice(actualPrefixTokenCount),
      text: text.slice(actualPrefixStrLength)
    };
  }

  /**
   * Decode an array of integer token ids
   * @param tokens array of integer token ids
   * @returns string decoded text
   */
  public decode(tokens: number[]): string {
    const decoded: number[] = [];
    for (const token of tokens) {
      let tokenBytes: number[] = [];
      const value = this.decoder?.get(token);
      if (value !== undefined) {
        tokenBytes = Array.from(value);
      } else {
        const specialTokenValue = this.specialTokensDecoder?.get(token);
        if (specialTokenValue !== undefined) {
          const bytes = this.textEncoder.encode(specialTokenValue);
          tokenBytes = Array.from(bytes.subarray(0, this.textEncoder.length));
        }
      }
      decoded.push(...tokenBytes);
    }

    return this.textDecoder.decode(new Uint8Array(decoded));
  }
}
